lxd/device/gpu_physical.go

```diff
@@ -98,10 +101,192 @@ func (d *gpuPhysical) Start() (*deviceConfig.RunConfig, error) {
 	return d.startContainer()
 }
 
+// startCDIDevices starts all the devices given in a CDI specification:
+// * `unix-char` (representing the card and non-card devices)
+// * `disk` (representing the mounts)).
+func (d *gpuPhysical) startCDIDevices(configDevices *cdi.ConfigDevices, hooks *cdi.Hooks, runConf *deviceConfig.RunConfig) error {
+	srcFDHandlers := make([]*os.File, 0)
+	defer func() {
+		for _, f := range srcFDHandlers {
+			_ = f.Close()
+		}
+	}()
+
+	for deviceName, conf := range configDevices.UnixCharDevs {
+		if conf["source"] == "" {
+			return fmt.Errorf("The source of the unix-char device %q is empty", deviceName)
+		}
+
+		_, major, minor, err := unixDeviceAttributes(conf["source"])
+		if err != nil {
+			return err
+		}
+
+		err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "unix", deviceName, conf, major, minor, conf["path"], false, runConf)
+		if err != nil {
+			return err
+		}
+	}
+
+	for deviceName, conf := range configDevices.DiskDevs {
+		srcPath := shared.HostPath(conf["source"])
+		destPath := conf["path"]
+		relativeDestPath := strings.TrimPrefix(destPath, "/")
+		ownerShift := deviceConfig.MountOwnerShiftNone
+		if shared.IsTrue(conf["shift"]) {
+			ownerShift = deviceConfig.MountOwnerShiftDynamic
+		}
+
+		options := []string{"bind"}
+		mntOptions := shared.SplitNTrimSpace(conf["raw.mount.options"], ",", -1, true)
+		devPath := filepath.Join(d.inst.DevicesPath(), filesystem.PathNameEncode(deviceJoinPath("disk", deviceName, relativeDestPath)))
+		fsName := "none"
+
+		fileInfo, err := os.Stat(srcPath)
+		if err != nil {
+			return fmt.Errorf("Failed accessing source path %q: %w", srcPath, err)
+		}
+
+		fileMode := fileInfo.Mode()
+		isFile := false
+		if !fileMode.IsDir() {
+			isFile = true
+		}
+
+		f, err := os.OpenFile(srcPath, unix.O_PATH|unix.O_CLOEXEC, 0)
```

Comment:
f doesnt appear to be closed, should this be a defer?

Reply:
@tomponline it is closed actually. If you look at the beginning of the function, you'll see:
```go
srcFDHandlers := make([]*os.File, 0)
defer func() {
		for _, f := range srcFDHandlers {
			_ = f.Close()
		}
}()
```

I did this because we can't have a `defer` statement in a loop and I need to keep `f` open until after the call to `DiskMount` at the end of the loop.

Reply:
Got you, thanks!